<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 5 Cluster Markers | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="Why do we need to do this? Single cell data helps to segregate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="Chapter 5 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="Why do we need to do this? Single cell data helps to segregate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 5 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="Why do we need to do this? Single cell data helps to segregate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.28/datatables.js"></script><link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">2</span> Schedule</a></li>
<li class="book-part">Seurat Preprocessing</li>
<li><a class="" href="preprocessing.html"><span class="header-section-number">3</span> Seurat PBMC3k Tutorial</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">4</span> Clustering</a></li>
<li><a class="active" href="clustermarkers.html"><span class="header-section-number">5</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">6</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">7</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">8</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">9</span> Data set integration with Harmony</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">10</span> Resources</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="clustermarkers" class="section level1" number="5">
<h1>
<span class="header-section-number">5</span> Cluster Markers<a class="anchor" aria-label="anchor" href="#clustermarkers"><i class="fas fa-link"></i></a>
</h1>
<div id="why-do-we-need-to-do-this" class="section level4 unnumbered rational">
<h4>Why do we need to do this?<a class="anchor" aria-label="anchor" href="#why-do-we-need-to-do-this"><i class="fas fa-link"></i></a>
</h4>
<p>Single cell data helps to segregate cell types. Use markers to identify cell types. warning: In this example the cell types/markers are well known and making this step easy, but in reality this step needs the experts curation.</p>
</div>
<div id="section-2" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-2"><i class="fas fa-link"></i></a></h4>
</div>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level2" number="5.1">
<h2>
<span class="header-section-number">5.1</span> Finding differentially expressed features (cluster biomarkers)<a class="anchor" aria-label="anchor" href="#finding-differentially-expressed-features-cluster-biomarkers"><i class="fas fa-link"></i></a>
</h2>
<p>Tutorial: <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial#finding-differentially-expressed-features-cluster-biomarkers" class="uri">https://satijalab.org/seurat/articles/pbmc3k_tutorial#finding-differentially-expressed-features-cluster-biomarkers</a></p>
</div>
<div id="use-markers-to-label-or-find-a-cluster" class="section level2" number="5.2">
<h2>
<span class="header-section-number">5.2</span> Use markers to label or find a cluster<a class="anchor" aria-label="anchor" href="#use-markers-to-label-or-find-a-cluster"><i class="fas fa-link"></i></a>
</h2>
<p>If you know markers for your cell types, use AddModuleScore to label them.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Naive_CD4_T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"IL7R"</span>, <span class="st">"CCR7"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AddModuleScore.html">AddModuleScore</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">genes_markers</span>, ctrl <span class="op">=</span> <span class="fl">5</span>, name <span class="op">=</span> <span class="st">"Naive_CD4_T"</span>,</span>
<span>    search <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co"># notice the name of the cluster has a 1 at the end</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "orig.ident"      "nCount_RNA"      "nFeature_RNA"   </span></span>
<span><span class="co">#&gt;  [4] "percent.mt"      "RNA_snn_res.0.5" "seurat_clusters"</span></span>
<span><span class="co">#&gt;  [7] "RNA_snn_res.0.1" "RNA_snn_res.0.2" "RNA_snn_res.0.3"</span></span>
<span><span class="co">#&gt; [10] "RNA_snn_res.0.4" "RNA_snn_res.0.6" "RNA_snn_res.0.7"</span></span>
<span><span class="co">#&gt; [13] "RNA_snn_res.0.8" "RNA_snn_res.0.9" "RNA_snn_res.1"  </span></span>
<span><span class="co">#&gt; [16] "RNA_snn_res.1.1" "RNA_snn_res.1.2" "RNA_snn_res.1.3"</span></span>
<span><span class="co">#&gt; [19] "RNA_snn_res.1.4" "RNA_snn_res.1.5" "RNA_snn_res.1.6"</span></span>
<span><span class="co">#&gt; [22] "RNA_snn_res.1.7" "RNA_snn_res.1.8" "RNA_snn_res.1.9"</span></span>
<span><span class="co">#&gt; [25] "RNA_snn_res.2"   "Naive_CD4_T1"</span></span>
<span></span>
<span><span class="co"># label that cell type</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">cell_label</span> <span class="op">=</span> <span class="cn">NA</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">cell_label</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">Naive_CD4_T1</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="st">"Naive_CD4_T"</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">=</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cell_label</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="co"># Using a custom colour scale </span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="st">"Naive_CD4_T1"</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, repel <span class="op">=</span> <span class="cn">TRUE</span>, <span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html">scale_colour_gradientn</a></span><span class="op">(</span>colours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"lightblue"</span>,<span class="st">"beige"</span>,<span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Scale for colour is already present.</span></span>
<span><span class="co">#&gt; Adding another scale for colour, which will replace the</span></span>
<span><span class="co">#&gt; existing scale.</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/unnamed-chunk-15-1.png" width="672"></div>
</div>
<div id="assigning-cell-type-identity-to-clusters" class="section level2" number="5.3">
<h2>
<span class="header-section-number">5.3</span> Assigning cell type identity to clusters<a class="anchor" aria-label="anchor" href="#assigning-cell-type-identity-to-clusters"><i class="fas fa-link"></i></a>
</h2>
<p>Tutorial: <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial#assigning-cell-type-identity-to-clusters" class="uri">https://satijalab.org/seurat/articles/pbmc3k_tutorial#assigning-cell-type-identity-to-clusters</a></p>
<p>This is presentend in a very simple way in the tutorial, an expanded approach is below.</p>
<p>First, it can avoid downstream hassles if you add a letter to your clusters so they are no longer just numbers. Here, we prefix a ‘c’ onto the cluster number, and store as a factor.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Sometimes you need to change Idents, so make sure your favourite cluster is stored in its own column too!</span></span>
<span><span class="co"># 5 =&gt; c5</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"c"</span>, <span class="va">pbmc</span><span class="op">$</span><span class="va">RNA_snn_res.0.5</span><span class="op">)</span>,   levels<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">'c'</span>, <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">RNA_snn_res.0.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "c0" "c1" "c2" "c3" "c4" "c5" "c6" "c7" "c8"</span></span></code></pre></div>
<p>We have clusters c0 to c8.</p>
<p>Now, its time to make a biologically-informed call of what cluster corresponds to what cell type. This step can take time! It often won’t be possible to confidently label all clusters. Some approaches we can use:</p>
<ul>
<li>Look at the cluster marker genes.</li>
<li>Look at known cell type markers - are they restricted to the cluster on the UMAP?</li>
<li>Use a celltype assignment tool (like singleR) to call celltypes against a reference dataset</li>
<li>Check the QC metrics on difficult to understand clusters - often cells with low counts will group together. These might be an ‘unknown’/‘poor quality’ cluster.</li>
<li>Check for weird clusters from just one sample. They might be spurious.</li>
<li>To separate similar clusters, you can subset your data and calculate cell markers within the subset.</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Template</span></span>
<span><span class="va">cluster_content</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  c0 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c1 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c2 <span class="op">=</span> <span class="st">""</span>,  </span>
<span>  c3 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c4 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c5 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c6 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c7 <span class="op">=</span> <span class="st">""</span>,</span>
<span>  c8 <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Finally, when you’ve some up with some labels - you can run store them in your object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster_content</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span></span>
<span>  c0 <span class="op">=</span> <span class="st">"Naive CD4+ T"</span>,</span>
<span>  c1 <span class="op">=</span> <span class="st">"CD14+ Mono"</span>,</span>
<span>  c2 <span class="op">=</span> <span class="st">"Memory CD4+"</span>,  </span>
<span>  c3 <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>  c4 <span class="op">=</span> <span class="st">"CD8+ T"</span>,</span>
<span>  c5 <span class="op">=</span> <span class="st">"Mono"</span>,</span>
<span>  c6 <span class="op">=</span> <span class="st">"NK"</span>,</span>
<span>  c7 <span class="op">=</span> <span class="st">"DC"</span>,</span>
<span>  c8 <span class="op">=</span> <span class="st">"Platelet"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># "c5" =&gt; "Mono" </span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>With a little extra code, we can keep the order of our celltypes, and create labels
that incorporate the cluster number. It can be useful to keep cluster labels during your analysis
(‘CD4+ T’ makes for terrible filenames!), especially where you have multiple clusters with the same celltype (e.g. multiple similar yet different ‘stroma’ celltypes are common).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># "c5" =&gt; "Mono" </span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">celltype</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">]</span><span class="op">)</span>, levels<span class="op">=</span><span class="va">cluster_content</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># c5 =&gt; c5: Mono</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">pretty_cluster_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">factor</a></span> <span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">]</span><span class="op">)</span>, <span class="st">": "</span>, <span class="va">cluster_content</span><span class="op">[</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cluster</span><span class="op">]</span><span class="op">)</span> , </span>
<span>  levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">cluster_content</span><span class="op">)</span>, <span class="st">": "</span>, <span class="va">cluster_content</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>What’s in our sample?</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">RNA_snn_res.0.5</span>, <span class="va">cluster</span>, <span class="va">celltype</span>, <span class="va">pretty_cluster_labels</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>num_cells<span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html">n</a></span><span class="op">(</span><span class="op">)</span>, .groups<span class="op">=</span><span class="st">"drop"</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">DT</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/DT/man/datatable.html">datatable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-830e81e5aaafc599ffb2" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-830e81e5aaafc599ffb2">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9"],["0","1","2","3","4","5","6","7","8"],["c0","c1","c2","c3","c4","c5","c6","c7","c8"],["Naive CD4+ T","CD14+ Mono","Memory CD4+","B","CD8+ T","Mono","NK","DC","Platelet"],["c0: Naive CD4+ T","c1: CD14+ Mono","c2: Memory CD4+","c3: B","c4: CD8+ T","c5: Mono","c6: NK","c7: DC","c8: Platelet"],[684,481,476,344,291,162,155,32,13]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>RNA_snn_res.0.5<\/th>\n      <th>cluster<\/th>\n      <th>celltype<\/th>\n      <th>pretty_cluster_labels<\/th>\n      <th>num_cells<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":5},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script><p>Can we figure out cell types automatically? Yes! There are methods like singleR but they are not perfect.</p>

</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="clustering.html"><span class="header-section-number">4</span> Clustering</a></div>
<div class="next"><a href="singler.html"><span class="header-section-number">6</span> SingleR</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#clustermarkers"><span class="header-section-number">5</span> Cluster Markers</a></li>
<li><a class="nav-link" href="#finding-differentially-expressed-features-cluster-biomarkers"><span class="header-section-number">5.1</span> Finding differentially expressed features (cluster biomarkers)</a></li>
<li><a class="nav-link" href="#use-markers-to-label-or-find-a-cluster"><span class="header-section-number">5.2</span> Use markers to label or find a cluster</a></li>
<li><a class="nav-link" href="#assigning-cell-type-identity-to-clusters"><span class="header-section-number">5.3</span> Assigning cell type identity to clusters</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/04-07-clustermarkers.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/04-07-clustermarkers.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2024-05-10.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
