<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 9 Cluster Markers | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="9.1 Finding differentially expressed features (cluster biomarkers) Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="Chapter 9 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="9.1 Finding differentially expressed features (cluster biomarkers) Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 9 Cluster Markers | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="9.1 Finding differentially expressed features (cluster biomarkers) Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">2</span> Schedule</a></li>
<li class="book-part">Content</li>
<li><a class="" href="seuratobject.html"><span class="header-section-number">3</span> The Seurat object</a></li>
<li class="book-part">Seurat PBMC3k Tutorial</li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li><a class="" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">8</span> Clustering</a></li>
<li><a class="active" href="cluster-markers.html"><span class="header-section-number">9</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">10</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">11</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">12</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">13</span> Data set integration with Harmony</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">14</span> Resources</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="cluster-markers" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Cluster Markers<a class="anchor" aria-label="anchor" href="#cluster-markers"><i class="fas fa-link"></i></a>
</h1>
<div id="finding-differentially-expressed-features-cluster-biomarkers" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Finding differentially expressed features (cluster biomarkers)<a class="anchor" aria-label="anchor" href="#finding-differentially-expressed-features-cluster-biomarkers"><i class="fas fa-link"></i></a>
</h2>
<p>Seurat can help you find markers that define clusters via differential expression. By default, it identifies positive and negative markers of a single cluster (specified in <code>ident.1</code>), compared to all other cells. <code><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers()</a></code> automates this process for all clusters, but you can also test groups of clusters vs. each other, or against all cells.</p>
<p>The <code>min.pct</code> argument requires a feature to be detected at a minimum percentage in either of the two groups of cells, and the thresh.test argument requires a feature to be differentially expressed (on average) by some amount between the two groups. You can set both of these to 0, but with a dramatic increase in time - since this will test a large number of features that are unlikely to be highly discriminatory. As another option to speed up these computations, <code>max.cells.per.ident</code> can be set. This will downsample each identity class to have no more cells than whatever this is set to. While there is generally going to be a loss in power, the speed increases can be significant and the most highly differentially expressed features will likely still rise to the top.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># find all markers of cluster 2</span></span>
<span><span class="va">cluster2.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">2</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster2.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;             p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">#&gt; IL32 2.892340e-90  1.2013522 0.947 0.465 3.966555e-86</span></span>
<span><span class="co">#&gt; LTB  1.060121e-86  1.2695776 0.981 0.643 1.453850e-82</span></span>
<span><span class="co">#&gt; CD3D 8.794641e-71  0.9389621 0.922 0.432 1.206097e-66</span></span>
<span><span class="co">#&gt; IL7R 3.516098e-68  1.1873213 0.750 0.326 4.821977e-64</span></span>
<span><span class="co">#&gt; LDHB 1.642480e-67  0.8969774 0.954 0.614 2.252497e-63</span></span>
<span><span class="co"># find all markers distinguishing cluster 5 from clusters 0 and 3</span></span>
<span><span class="va">cluster5.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">5</span>, ident.2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">3</span><span class="op">)</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">cluster5.markers</span>, n <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt;                       p_val avg_log2FC pct.1 pct.2</span></span>
<span><span class="co">#&gt; FCGR3A        8.246578e-205   4.261495 0.975 0.040</span></span>
<span><span class="co">#&gt; IFITM3        1.677613e-195   3.879339 0.975 0.049</span></span>
<span><span class="co">#&gt; CFD           2.401156e-193   3.405492 0.938 0.038</span></span>
<span><span class="co">#&gt; CD68          2.900384e-191   3.020484 0.926 0.035</span></span>
<span><span class="co">#&gt; RP11-290F20.3 2.513244e-186   2.720057 0.840 0.017</span></span>
<span><span class="co">#&gt;                   p_val_adj</span></span>
<span><span class="co">#&gt; FCGR3A        1.130936e-200</span></span>
<span><span class="co">#&gt; IFITM3        2.300678e-191</span></span>
<span><span class="co">#&gt; CFD           3.292945e-189</span></span>
<span><span class="co">#&gt; CD68          3.977587e-187</span></span>
<span><span class="co">#&gt; RP11-290F20.3 3.446663e-182</span></span>
<span><span class="co"># find markers for every cluster compared to all remaining cells, report only the positive ones</span></span>
<span><span class="va">pbmc.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindAllMarkers.html">FindAllMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span>, min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span><span class="co">#&gt; Calculating cluster 0</span></span>
<span><span class="co">#&gt; Calculating cluster 1</span></span>
<span><span class="co">#&gt; Calculating cluster 2</span></span>
<span><span class="co">#&gt; Calculating cluster 3</span></span>
<span><span class="co">#&gt; Calculating cluster 4</span></span>
<span><span class="co">#&gt; Calculating cluster 5</span></span>
<span><span class="co">#&gt; Calculating cluster 6</span></span>
<span><span class="co">#&gt; Calculating cluster 7</span></span>
<span><span class="co">#&gt; Calculating cluster 8</span></span>
<span><span class="va">pbmc.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice_max</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">2</span>, order_by <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="co">#&gt; # A tibble: 18 × 7</span></span>
<span><span class="co">#&gt; # Groups:   cluster [9]</span></span>
<span><span class="co">#&gt;        p_val avg_log2FC pct.1 pct.2 p_val_adj cluster gene  </span></span>
<span><span class="co">#&gt;        &lt;dbl&gt;      &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt; </span></span>
<span><span class="co">#&gt;  1 9.57e- 88       1.36 0.447 0.108 1.31e- 83 0       CCR7  </span></span>
<span><span class="co">#&gt;  2 3.75e-112       1.09 0.912 0.592 5.14e-108 0       LDHB  </span></span>
<span><span class="co">#&gt;  3 0               5.57 0.996 0.215 0         1       S100A9</span></span>
<span><span class="co">#&gt;  4 0               5.48 0.975 0.121 0         1       S100A8</span></span>
<span><span class="co">#&gt;  5 1.06e- 86       1.27 0.981 0.643 1.45e- 82 2       LTB   </span></span>
<span><span class="co">#&gt;  6 2.97e- 58       1.23 0.42  0.111 4.07e- 54 2       AQP3  </span></span>
<span><span class="co">#&gt;  7 0               4.31 0.936 0.041 0         3       CD79A </span></span>
<span><span class="co">#&gt;  8 9.48e-271       3.59 0.622 0.022 1.30e-266 3       TCL1A </span></span>
<span><span class="co">#&gt;  9 5.61e-202       3.10 0.983 0.234 7.70e-198 4       CCL5  </span></span>
<span><span class="co">#&gt; 10 7.25e-165       3.00 0.577 0.055 9.95e-161 4       GZMK  </span></span>
<span><span class="co">#&gt; 11 3.51e-184       3.31 0.975 0.134 4.82e-180 5       FCGR3A</span></span>
<span><span class="co">#&gt; 12 2.03e-125       3.09 1     0.315 2.78e-121 5       LST1  </span></span>
<span><span class="co">#&gt; 13 3.13e-191       5.32 0.961 0.131 4.30e-187 6       GNLY  </span></span>
<span><span class="co">#&gt; 14 7.95e-269       4.83 0.961 0.068 1.09e-264 6       GZMB  </span></span>
<span><span class="co">#&gt; 15 1.48e-220       3.87 0.812 0.011 2.03e-216 7       FCER1A</span></span>
<span><span class="co">#&gt; 16 1.67e- 21       2.87 1     0.513 2.28e- 17 7       HLA-D…</span></span>
<span><span class="co">#&gt; 17 1.92e-102       8.59 1     0.024 2.63e- 98 8       PPBP  </span></span>
<span><span class="co">#&gt; 18 9.25e-186       7.29 1     0.011 1.27e-181 8       PF4</span></span></code></pre></div>
<p>Seurat has several tests for differential expression which can be set with the test.use parameter (see our <a href="de_vignette.html">DE vignette</a> for details). For example, the ROC test returns the ‘classification power’ <code>abs(AUC-0.5)*2</code> for any individual marker, ranging from 0 = random to 1 = perfect.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cluster0.markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindMarkers.html">FindMarkers</a></span><span class="op">(</span><span class="va">pbmc</span>, ident.1 <span class="op">=</span> <span class="fl">0</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, test.use <span class="op">=</span> <span class="st">"roc"</span>, only.pos <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>We include several tools for visualizing marker expression. <code><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot()</a></code> (shows expression probability distributions across clusters), and <code><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot()</a></code> (visualizes feature expression on a tSNE or PCA plot) are our most commonly used visualizations. We also suggest exploring <code><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot()</a></code>, <code><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter()</a></code>, and <code><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot()</a></code> as additional methods to view your dataset.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"CD79A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/markerplots-1.png" width="1440"></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># you can plot raw counts as well</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"NKG7"</span>, <span class="st">"PF4"</span><span class="op">)</span>, slot <span class="op">=</span> <span class="st">'counts'</span>, log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/markerplots-2.png" width="1440"></div>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<img src="scRNAseqInR_Doco_files/figure-html/markerplots-3.png" width="1440"><details><summary><strong>Other useful plots</strong>
</summary><p>These are ridgeplots, cell scatter plots and dotplots. Replace <code>FeaturePlot</code> with the other functions.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html">RidgePlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0236</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0859</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.126</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0337</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0659</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0582</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.309</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0156</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 0.0366</span></span></code></pre></div>
<p><img src="scRNAseqInR_Doco_files/figure-html/ridgeplots-1.png" width="1440">
For CellScatter plots, will need the cell id of the cells you want to look at. You can get this from the cell metadata (<code>pbmc@meta.data</code>).</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span> <span class="va">pbmc</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">)</span></span>
<span><span class="co">#&gt;                  orig.ident nCount_RNA nFeature_RNA</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1     pbmc3k       2419          779</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1     pbmc3k       4903         1352</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1     pbmc3k       3147         1129</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1     pbmc3k       2639          960</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1     pbmc3k        980          521</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1     pbmc3k       2163          781</span></span>
<span><span class="co">#&gt;                  percent.mt RNA_snn_res.0.5 seurat_clusters</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1  3.0177759               2               6</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1  3.7935958               3               1</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1  0.8897363               2               0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1  1.7430845               1               5</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1  1.2244898               6               8</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1  1.6643551               2               0</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.1 RNA_snn_res.0.2</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               0               0</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               3               3</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               0               0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               1               1</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               2               2</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               0               0</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.3 RNA_snn_res.0.4</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               0               2</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               3               3</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               0               2</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               1               1</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               2               6</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               0               2</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.6 RNA_snn_res.0.7</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               1               1</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               3               3</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               1               1</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               2               2</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               6               6</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               1               1</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.0.8 RNA_snn_res.0.9</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               6               1</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               2               2</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               1               1</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               4               4</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               8               7</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               1               1</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1 RNA_snn_res.1.1</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1             6               6</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1             2               2</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1             1               1</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1             4               4</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1             8               8</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1             1               1</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.2 RNA_snn_res.1.3</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               6               8</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               2               2</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               1               0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               4               5</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               8               9</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               1               0</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.4 RNA_snn_res.1.5</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               8               9</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               2               2</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               0               1</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               5               4</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               9               8</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               0               1</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.6 RNA_snn_res.1.7</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               8               8</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               1               1</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               0               0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               3               3</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               7               7</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               0               0</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.1.8 RNA_snn_res.1.9</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1               7               6</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1               1               1</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1               0               0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1               3               3</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1               8               8</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1               0               0</span></span>
<span><span class="co">#&gt;                  RNA_snn_res.2</span></span>
<span><span class="co">#&gt; AAACATACAACCAC-1             6</span></span>
<span><span class="co">#&gt; AAACATTGAGCTAC-1             1</span></span>
<span><span class="co">#&gt; AAACATTGATCAGC-1             0</span></span>
<span><span class="co">#&gt; AAACCGTGCTTCCG-1             5</span></span>
<span><span class="co">#&gt; AAACCGTGTATGCG-1             8</span></span>
<span><span class="co">#&gt; AAACGCACTGGTAC-1             0</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/CellScatter.html">CellScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, cell1 <span class="op">=</span> <span class="st">"AAACATACAACCAC-1"</span>, cell2 <span class="op">=</span> <span class="st">"AAACATTGAGCTAC-1"</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/unnamed-chunk-29-1.png" width="672"></div>
<p>DotPlots</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DotPlot.html">DotPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MS4A1"</span>, <span class="st">"GNLY"</span>, <span class="st">"CD3E"</span>, <span class="st">"CD14"</span>, <span class="st">"FCER1A"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYZ"</span>, <span class="st">"PPBP"</span>, <span class="st">"CD8A"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/unnamed-chunk-30-1.png" width="672"></div>
<p>Which plots do you prefer? Discuss.</p>
</details><p><code><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap()</a></code> generates an expression heatmap for given cells and features. In this case, we are plotting the top 20 markers (or all markers if less than 20) for each cluster.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top10</span> <span class="op">&lt;-</span> <span class="va">pbmc.markers</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">cluster</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/top_n.html">top_n</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">10</span>, wt <span class="op">=</span> <span class="va">avg_log2FC</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DoHeatmap.html">DoHeatmap</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="va">top10</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/clusterHeatmap-1.png" width="1440"></div>
<hr>
</div>
<div id="assigning-cell-type-identity-to-clusters" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Assigning cell type identity to clusters<a class="anchor" aria-label="anchor" href="#assigning-cell-type-identity-to-clusters"><i class="fas fa-link"></i></a>
</h2>
<p>Fortunately in the case of this dataset, we can use canonical markers to easily match the unbiased clustering to known cell types:</p>
<div class="inline-table"><table class="table table-sm">
<thead><tr class="header">
<th>Cluster ID</th>
<th>Markers</th>
<th>Cell Type</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>0</td>
<td>IL7R, CCR7</td>
<td>Naive CD4+ T</td>
</tr>
<tr class="even">
<td>1</td>
<td>CD14, LYZ</td>
<td>CD14+ Mono</td>
</tr>
<tr class="odd">
<td>2</td>
<td>IL7R, S100A4</td>
<td>Memory CD4+</td>
</tr>
<tr class="even">
<td>3</td>
<td>MS4A1</td>
<td>B</td>
</tr>
<tr class="odd">
<td>4</td>
<td>CD8A</td>
<td>CD8+ T</td>
</tr>
<tr class="even">
<td>5</td>
<td>FCGR3A, MS4A7</td>
<td>FCGR3A+ Mono</td>
</tr>
<tr class="odd">
<td>6</td>
<td>GNLY, NKG7</td>
<td>NK</td>
</tr>
<tr class="even">
<td>7</td>
<td>FCER1A, CST3</td>
<td>DC</td>
</tr>
<tr class="odd">
<td>8</td>
<td>PPBP</td>
<td>Platelet</td>
</tr>
</tbody>
</table></div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Naive CD4 T"</span>, <span class="st">"CD14+ Mono"</span>, <span class="st">"Memory CD4 T"</span>, <span class="st">"B"</span>, <span class="st">"CD8 T"</span>, <span class="st">"FCGR3A+ Mono"</span>, <span class="st">"NK"</span>, <span class="st">"DC"</span>, <span class="st">"Platelet"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/Idents.html">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">new.cluster.ids</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">'umap'</span>, label <span class="op">=</span> <span class="cn">TRUE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/SeuratTheme.html">NoLegend</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/labelplot-1.png" width="864"></div>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">pbmc</span>, file <span class="op">=</span> <span class="st">"pbmc3k_final.rds"</span><span class="op">)</span></span></code></pre></div>

</div>
</div>



  <div class="chapter-nav">
<div class="prev"><a href="clustering.html"><span class="header-section-number">8</span> Clustering</a></div>
<div class="next"><a href="singler.html"><span class="header-section-number">10</span> SingleR</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#cluster-markers"><span class="header-section-number">9</span> Cluster Markers</a></li>
<li><a class="nav-link" href="#finding-differentially-expressed-features-cluster-biomarkers"><span class="header-section-number">9.1</span> Finding differentially expressed features (cluster biomarkers)</a></li>
<li><a class="nav-link" href="#assigning-cell-type-identity-to-clusters"><span class="header-section-number">9.2</span> Assigning cell type identity to clusters</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/04-07-clustermarkers.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/04-07-clustermarkers.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2023-09-11.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
