<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 5 QC Filtering | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="Tutorial: https://satijalab.org/seurat/articles/pbmc3k_tutorial#qc-and-selecting-cells-for-further-analysis     5.1 QC and selecting cells for further analysis  Why do we need to do this? Low...">
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="Chapter 5 QC Filtering | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="Tutorial: https://satijalab.org/seurat/articles/pbmc3k_tutorial#qc-and-selecting-cells-for-further-analysis     5.1 QC and selecting cells for further analysis  Why do we need to do this? Low...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 5 QC Filtering | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="Tutorial: https://satijalab.org/seurat/articles/pbmc3k_tutorial#qc-and-selecting-cells-for-further-analysis     5.1 QC and selecting cells for further analysis  Why do we need to do this? Low...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">2</span> Schedule</a></li>
<li class="book-part">Content</li>
<li><a class="" href="seuratobject.html"><span class="header-section-number">3</span> The Seurat object</a></li>
<li class="book-part">Seurat PBMC3k Tutorial</li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li><a class="active" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="dimensionality-reduction.html"><span class="header-section-number">8</span> Dimensionality reduction</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">9</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">10</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">11</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">12</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">13</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">14</span> Data set integration with Harmony</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">15</span> Resources</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="qc" class="section level1" number="5">
<h1>
<span class="header-section-number">5</span> QC Filtering<a class="anchor" aria-label="anchor" href="#qc"><i class="fas fa-link"></i></a>
</h1>
<p>Tutorial: <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial#qc-and-selecting-cells-for-further-analysis" class="uri">https://satijalab.org/seurat/articles/pbmc3k_tutorial#qc-and-selecting-cells-for-further-analysis</a></p>
<!-- Discuss counts per cell/gene and make plots -->
<!-- Discuss that there's no one threshold -->
<!-- everyone pick a threshold and go filter -->
<!-- check numbers of cells. -->
<!-- save your object -->
<!-- (other qc metrics = Mt gene content, cell cycle asignment, low seq diversity e.t.c) -->
<div id="qc-and-selecting-cells-for-further-analysis" class="section level2" number="5.1">
<h2>
<span class="header-section-number">5.1</span> QC and selecting cells for further analysis<a class="anchor" aria-label="anchor" href="#qc-and-selecting-cells-for-further-analysis"><i class="fas fa-link"></i></a>
</h2>
<div id="why-do-we-need-to-do-this" class="section level4 unnumbered rational">
<h4>Why do we need to do this?<a class="anchor" aria-label="anchor" href="#why-do-we-need-to-do-this"><i class="fas fa-link"></i></a>
</h4>
<p>Low quality cells can add noise to your results leading you to the wrong biological conclusions. Using only good quality cells helps you to avoid this. Reduce noise in the data by filtering out low quality cells such as dying or stressed cells (high mitochondrial expression) and cells with few features that can reflect empty droplets.</p>
</div>
<div id="section" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section"><i class="fas fa-link"></i></a></h4>
</div>
<div id="challenge-the-meta.data-slot-in-the-seurat-object" class="section level4 unnumbered challenge">
<h4>Challenge: The meta.data slot in the Seurat object<a class="anchor" aria-label="anchor" href="#challenge-the-meta.data-slot-in-the-seurat-object"><i class="fas fa-link"></i></a>
</h4>
<p>Where are QC metrics stored in Seurat?</p>
<ul>
<li>The number of unique genes and total molecules are automatically calculated during <code><a href="https://rdrr.io/pkg/SeuratObject/man/CreateSeuratObject.html">CreateSeuratObject()</a></code>
<ul>
<li>You can find them stored in the object meta data</li>
</ul>
</li>
</ul>
<ol style="list-style-type: decimal">
<li><p>What do you notice has changed within the <code>meta.data</code> table now that we have calculated mitochondrial gene proportion?</p></li>
<li><p>Imagine that this is the first of
several samples in our experiment. Add a <code>samplename</code> column to to the <code>meta.data</code> table.</p></li>
</ol>
</div>
<div id="section-1" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-1"><i class="fas fa-link"></i></a></h4>
</div>
<div id="challenge-filter-the-cells" class="section level4 unnumbered challenge">
<h4>Challenge: Filter the cells<a class="anchor" aria-label="anchor" href="#challenge-filter-the-cells"><i class="fas fa-link"></i></a>
</h4>
<p>Apply the filtering thresolds defined above.</p>
<ul>
<li>How many cells survived filtering?</li>
</ul>
<p>The PBMC3k dataset weâ€™re working with in this tutorial is quite old. There are a number of other example datasets available from the 10X website, including <a href="https://www.10xgenomics.com/resources/datasets/10k-human-pbmcs-3-v3-1-chromium-x-with-intronic-reads-3-1-high">this one</a> - published in 2022, sequencing 10k PBMCs with a newer chemistry and counting method.</p>
<ul>
<li>What thresholds would you chose to apply to this modern dataset?</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc10k_unfiltered</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">"data/10k_PBMC_v3.1ChromiumX_Intronic.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc10k_unfiltered</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
</div>
<div id="section-2" class="section level4 unnumbered">
<h4 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-2"><i class="fas fa-link"></i></a></h4>
</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="load.html"><span class="header-section-number">4</span> Load data</a></div>
<div class="next"><a href="norm.html"><span class="header-section-number">6</span> Normalisation</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#qc"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="nav-link" href="#qc-and-selecting-cells-for-further-analysis"><span class="header-section-number">5.1</span> QC and selecting cells for further analysis</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/04-02-seurattutorial.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/04-02-seurattutorial.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2024-05-02.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
