<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 7 QC Filtering | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and...">
<meta name="generator" content="bookdown 0.27 with bs4_book()">
<meta property="og:title" content="Chapter 7 QC Filtering | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 7 QC Filtering | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="welcome.html"><span class="header-section-number">2</span> Welcome</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">3</span> Schedule</a></li>
<li class="book-part">Content</li>
<li><a class="" href="overview.html"><span class="header-section-number">4</span> Overview</a></li>
<li><a class="" href="seuratobject.html"><span class="header-section-number">5</span> Using a seurat object</a></li>
<li class="book-part">Seurat PBMC3k Tutorial</li>
<li><a class="" href="load.html"><span class="header-section-number">6</span> Load data</a></li>
<li><a class="active" href="qc.html"><span class="header-section-number">7</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">8</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">9</span> PCAs and UMAPs</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">10</span> Clustering</a></li>
<li><a class="" href="cluster-markers.html"><span class="header-section-number">11</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">12</span> SingleR</a></li>
<li><a class="" href="de.html"><span class="header-section-number">13</span> DE</a></li>
<li class="book-part">End</li>
<li><a class="" href="finishing-up.html"><span class="header-section-number">14</span> Finishing up</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">15</span> Resources</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="qc" class="section level1" number="7">
<h1>
<span class="header-section-number">7</span> QC Filtering<a class="anchor" aria-label="anchor" href="#qc"><i class="fas fa-link"></i></a>
</h1>
<!-- Discuss counts per cell/gene and make plots -->
<!-- Discuss that there's no one threshold -->
<!-- everyone pick a threshold and go filter -->
<!-- check numbers of cells. -->
<!-- save your object -->
<!-- (other qc metrics = Mt gene content, cell cycle asignment, low seq diversity e.t.c) -->
<p>The steps below encompass the standard pre-processing workflow for scRNA-seq data in Seurat. These represent the selection and filtration of cells based on QC metrics, data normalization and scaling, and the detection of highly variable features.</p>
<div id="qc-and-selecting-cells-for-further-analysis" class="section level2" number="7.1">
<h2>
<span class="header-section-number">7.1</span> QC and selecting cells for further analysis<a class="anchor" aria-label="anchor" href="#qc-and-selecting-cells-for-further-analysis"><i class="fas fa-link"></i></a>
</h2>
<p>Seurat allows you to easily explore QC metrics and filter cells based on any user-defined criteria. A few QC metrics <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4758103/">commonly used</a> by the community include</p>
<ul>
<li>The number of unique genes detected in each cell.
<ul>
<li>Low-quality cells or empty droplets will often have very few genes</li>
<li>Cell doublets or multiplets may exhibit an aberrantly high gene count</li>
</ul>
</li>
<li>Similarly, the total number of molecules detected within a cell (correlates strongly with unique genes)</li>
<li>The percentage of reads that map to the mitochondrial genome
<ul>
<li>Low-quality / dying cells often exhibit extensive mitochondrial contamination</li>
<li>We calculate mitochondrial QC metrics with the <code><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet()</a></code> function, which calculates the percentage of counts originating from a set of features</li>
<li>We use the set of all genes starting with <code>MT-</code> as a set of mitochondrial genes</li>
</ul>
</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The $ operator can add columns to object metadata. </span></span>
<span><span class="co"># This is a great place to stash QC stats</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">percent.mt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span></span></code></pre></div>
<div id="challenge-the-meta.data-slot-in-the-seurat-object" class="section level4 unnumbered rmdtip">
<h4>Challenge: The meta.data slot in the Seurat object<a class="anchor" aria-label="anchor" href="#challenge-the-meta.data-slot-in-the-seurat-object"><i class="fas fa-link"></i></a>
</h4>
<p>Where are QC metrics stored in Seurat?</p>
<ul>
<li>The number of unique genes and total molecules are automatically calculated during <code><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html">CreateSeuratObject()</a></code>
<ul>
<li>You can find them stored in the object meta data</li>
</ul>
</li>
</ul>
<p>What do you notice has changed within the <code>meta.data</code> table now that we have calculated mitochondrial gene proportion?</p>
<p>Could we add more data into the <code>meta.data</code> table?</p>
</div>
<div id="section" class="section level3 unnumbered">
<h3 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section"><i class="fas fa-link"></i></a></h3>
<p>In the example below, we visualize QC metrics, and use these to filter cells.</p>
<ul>
<li>We filter cells that have unique feature counts over 2,500 or less than 200</li>
<li>We filter cells that have &gt;5% mitochondrial counts</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#Visualize QC metrics as a violin plot</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"nFeature_RNA"</span>, <span class="st">"nCount_RNA"</span>, <span class="st">"percent.mt"</span><span class="op">)</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/qc2-1.png" width="1248"></div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># FeatureScatter is typically used to visualize feature-feature relationships, </span></span>
<span><span class="co"># but can be used for anything calculated by the object, </span></span>
<span><span class="co"># i.e. columns in object metadata, PC scores etc.</span></span>
<span><span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span> </span>
<span><span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span> </span>
<span><span class="va">plot1</span> <span class="op">+</span> <span class="va">plot2</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/qc2-2.png" width="1248"></div>
<p>Lets look at the number of features (genes) to the percent mitochondrial genes plot.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nFeature_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span> </span>
<span><span class="va">plot3</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/unnamed-chunk-16-1.png" width="672"></div>
<div id="challenge-ribosomal-gene-expression-as-a-qc-metric" class="section level4 unnumbered rmdtip">
<h4>Challenge: Ribosomal gene expression as a QC metric<a class="anchor" aria-label="anchor" href="#challenge-ribosomal-gene-expression-as-a-qc-metric"><i class="fas fa-link"></i></a>
</h4>
<p>Ribosomal gene expression could be another factor to look into your cells within your experiment.</p>
<p>Create more columns of metadata using <code>PercentageFeatureSet</code> function, this time search for ribosomal genes. We can calculate the percentage for the large subunit (RPL) and small subunit (RPS) ribosomal genes.</p>
<p>Use <code>FeatureScatter</code> to plot combinations of metrics available in metadata. How is the mitochondrial gene percentage related to the ribosomal gene percentage? What can you see? Discuss in break out.</p>
<details><summary><strong>Code for challenge</strong>
</summary><p>Create new meta.data columns to contain percentages of the large and small ribosomal genes.</p>
<p>Then plot a scatter plot with this new data. You should find that the large and small ribosomal subunit genes are correlated within cell.</p>
<p>What about with mitochondria and gene, feature counts?</p>
These are the cells you may want to exclude.
</details><details><summary><strong>Advanced Challenge</strong>
</summary>
Highlight cells with very low percentage of ribosomal genes, create a new column in the meta.data table and with <code>FeatureScatter</code> make a plot of the RNA count and mitochondrial percentage with the cells with very low ribosomal gene perentage.
</details>
</div>
</div>
<div id="section-1" class="section level3 unnumbered">
<h3 class="unnumbered"><a class="anchor" aria-label="anchor" href="#section-1"><i class="fas fa-link"></i></a></h3>
<p>Okay we are happy with our thresholds for mitochondrial percentage in cells, lets apply them and subset our data. This will remove the cells we think are of poor quality.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre></div>
<p>Lets replot the feature scatters and see what they look like.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">plot5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"percent.mt"</span><span class="op">)</span> </span>
<span><span class="va">plot6</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FeatureScatter.html">FeatureScatter</a></span><span class="op">(</span><span class="va">pbmc</span>, feature1 <span class="op">=</span> <span class="st">"nCount_RNA"</span>, feature2 <span class="op">=</span> <span class="st">"nFeature_RNA"</span><span class="op">)</span> </span>
<span><span class="va">plot5</span> <span class="op">+</span> <span class="va">plot6</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/qc2_sidebar-1.png" width="1248"></div>

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="load.html"><span class="header-section-number">6</span> Load data</a></div>
<div class="next"><a href="norm.html"><span class="header-section-number">8</span> Normalisation</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#qc"><span class="header-section-number">7</span> QC Filtering</a></li>
<li>
<a class="nav-link" href="#qc-and-selecting-cells-for-further-analysis"><span class="header-section-number">7.1</span> QC and selecting cells for further analysis</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#section"></a></li>
<li><a class="nav-link" href="#section-1"></a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/04-03-qc.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/04-03-qc.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2022-07-17.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
