<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 13 Resources | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content="Useful resources for next steps.  13.1 Help and fruther Resources  Seurat Vignettes https://satijalab.org/seurat/index.html There are a good many Seurat vigettes for different aspects of the...">
<meta name="generator" content="bookdown 0.27 with bs4_book()">
<meta property="og:title" content="Chapter 13 Resources | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content="Useful resources for next steps.  13.1 Help and fruther Resources  Seurat Vignettes https://satijalab.org/seurat/index.html There are a good many Seurat vigettes for different aspects of the...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 13 Resources | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content="Useful resources for next steps.  13.1 Help and fruther Resources  Seurat Vignettes https://satijalab.org/seurat/index.html There are a good many Seurat vigettes for different aspects of the...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.3.1/transition.js"></script><script src="libs/bs3compat-0.3.1/tabs.js"></script><script src="libs/bs3compat-0.3.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.20/datatables.js"></script><link href="libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">2</span> Schedule</a></li>
<li class="book-part">Content</li>
<li><a class="" href="seuratobject.html"><span class="header-section-number">3</span> The Seurat object</a></li>
<li class="book-part">Seurat PBMC3k Tutorial</li>
<li><a class="" href="load.html"><span class="header-section-number">4</span> Load data</a></li>
<li><a class="" href="qc.html"><span class="header-section-number">5</span> QC Filtering</a></li>
<li><a class="" href="norm.html"><span class="header-section-number">6</span> Normalisation</a></li>
<li><a class="" href="reducedims.html"><span class="header-section-number">7</span> PCAs and UMAPs</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">8</span> Clustering</a></li>
<li><a class="" href="cluster-markers.html"><span class="header-section-number">9</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="" href="singler.html"><span class="header-section-number">10</span> SingleR</a></li>
<li><a class="" href="de.html"><span class="header-section-number">11</span> Differential Expression</a></li>
<li class="book-part">End</li>
<li><a class="" href="finishing-up.html"><span class="header-section-number">12</span> Finishing up</a></li>
<li><a class="active" href="resources.html"><span class="header-section-number">13</span> Resources</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="resources" class="section level1" number="13">
<h1>
<span class="header-section-number">13</span> Resources<a class="anchor" aria-label="anchor" href="#resources"><i class="fas fa-link"></i></a>
</h1>
<p>Useful resources for next steps.</p>
<!-- There's alot more we could add here!, Cell type, python AnnData information, Galaxy informatio, trajectory inference inof ...-->
<div id="help-and-fruther-resources" class="section level2" number="13.1">
<h2>
<span class="header-section-number">13.1</span> Help and fruther Resources<a class="anchor" aria-label="anchor" href="#help-and-fruther-resources"><i class="fas fa-link"></i></a>
</h2>
<div id="seurat-vignettes" class="section level3 unnumbered">
<h3>Seurat Vignettes<a class="anchor" aria-label="anchor" href="#seurat-vignettes"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://satijalab.org/seurat/index.html" class="uri">https://satijalab.org/seurat/index.html</a></p>
<p>There are a good many Seurat vigettes for different aspects of the Seurat package. E.g.</p>
<p><em><a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html">Guided Clustering tutorial</a> : We’ve just worked through this
</em><a href="https://satijalab.org/seurat/archive/v3.1/de_vignette.html">Differential expression</a> : An Exploration of differential expression methods within Seurat
*<a href="https://satijalab.org/seurat/articles/integration_introduction.html">Data integration</a> : Seurat’s data integration is a popular method to combine different datasets into one joint analysis.</p>
</div>
<div id="seurat-cheatsheet" class="section level3 unnumbered">
<h3>Seurat Cheatsheet<a class="anchor" aria-label="anchor" href="#seurat-cheatsheet"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://satijalab.org/seurat/articles/essential_commands.html" class="uri">https://satijalab.org/seurat/articles/essential_commands.html</a></p>
<p>A useful resource for asking; How can I do ‘X’ with my seurat object?</p>
</div>
<div id="osca" class="section level3 unnumbered">
<h3>OSCA<a class="anchor" aria-label="anchor" href="#osca"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://bioconductor.org/books/release/OSCA/" class="uri">https://bioconductor.org/books/release/OSCA/</a></p>
<p>An comprehensive resource for analysis approaches for single cell data.
This uses the SingleCellExperiment bioconductor ecosystem, but alot of the same principle still apply.</p>
<p>This includes a good discussion of useing <a href="http://bioconductor.org/books/3.15/OSCA.multisample/multi-sample-comparisons.html#creating-pseudo-bulk-samples">pseudobulk approaches</a>, worth checking out for differential expression analyses.</p>
</div>
<div id="mbp-training-reading-list" class="section level3 unnumbered">
<h3>MBP training Reading list<a class="anchor" aria-label="anchor" href="#mbp-training-reading-list"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://monashbioinformaticsplatform.github.io/Single-Cell-Workshop/" class="uri">https://monashbioinformaticsplatform.github.io/Single-Cell-Workshop/</a></p>
<p>A workshop page for a previous workshop (upon which this one is based) run by Monash Bioinformatics Platform - down the bottom there
is an extensive list of useful single cell links and resources.</p>
</div>
<div id="biocommons-single-cell-omics" class="section level3 unnumbered">
<h3>Biocommons Single Cell Omics<a class="anchor" aria-label="anchor" href="#biocommons-single-cell-omics"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://www.biocommons.org.au/single-cell-omics" class="uri">https://www.biocommons.org.au/single-cell-omics</a></p>
<p>Join the single cell omics community resources being setup by biocommons.</p>
<hr>
</div>
</div>
<div id="data" class="section level2" number="13.2">
<h2>
<span class="header-section-number">13.2</span> Data<a class="anchor" aria-label="anchor" href="#data"><i class="fas fa-link"></i></a>
</h2>
<div id="demo-10x-data" class="section level3 unnumbered">
<h3>Demo 10X data<a class="anchor" aria-label="anchor" href="#demo-10x-data"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://www.10xgenomics.com/resources/datasets" class="uri">https://www.10xgenomics.com/resources/datasets</a></p>
<p>10X genomics have quite a few example datasets availble for download (including PBMC3k).
This is a useful resource if you want to see what the ‘raw’ data looks like for a particular technology.</p>
</div>
<div id="geo" class="section level3 unnumbered">
<h3>GEO<a class="anchor" aria-label="anchor" href="#geo"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/" class="uri">https://www.ncbi.nlm.nih.gov/geo/</a></p>
<p>Many papers publish their raw single cell data in GEO. Formats vary, but often you can find the counts matrix.</p>
</div>
<div id="seurat-data" class="section level3 unnumbered">
<h3>Seurat data<a class="anchor" aria-label="anchor" href="#seurat-data"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://github.com/satijalab/seurat-data" class="uri">https://github.com/satijalab/seurat-data</a></p>
<p>Package for obtaining a few datasets as seurat objects.</p>
<!-- ### scRNAseq {-} -->
<!-- https://bioconductor.org/packages/release/data/experiment/html/scRNAseq.html -->
<!-- Preloaded sce data. -->
<hr>
</div>
</div>
<div id="analysis-tools" class="section level2" number="13.3">
<h2>
<span class="header-section-number">13.3</span> Analysis Tools<a class="anchor" aria-label="anchor" href="#analysis-tools"><i class="fas fa-link"></i></a>
</h2>
<p>A handful of the many tools that might be worth checking out for next steps.</p>
<div id="cyclone" class="section level3 unnumbered">
<h3>Cyclone<a class="anchor" aria-label="anchor" href="#cyclone"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://pubmed.ncbi.nlm.nih.gov/26142758/" class="uri">https://pubmed.ncbi.nlm.nih.gov/26142758/</a></p>
<p>Part of the scran package, cyclone is a(nother) method for determining cell phase.
<a href="https://rdrr.io/bioc/scran/man/cyclone.html">Doco</a></p>
</div>
<div id="harmony" class="section level3 unnumbered">
<h3>Harmony<a class="anchor" aria-label="anchor" href="#harmony"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://portals.broadinstitute.org/harmony/articles/quickstart.html" class="uri">https://portals.broadinstitute.org/harmony/articles/quickstart.html</a></p>
<p>Method for integration of multiple single cell datasets.</p>
</div>
<div id="singler-1" class="section level3 unnumbered">
<h3>SingleR<a class="anchor" aria-label="anchor" href="#singler-1"><i class="fas fa-link"></i></a>
</h3>
<p><a href="http://bioconductor.org/books/release/SingleRBook/" class="uri">http://bioconductor.org/books/release/SingleRBook/</a></p>
<p>There is extensive documentation for the singleR package in the ‘singleR’ book.</p>
</div>
<div id="scrublet" class="section level3 unnumbered">
<h3>Scrublet<a class="anchor" aria-label="anchor" href="#scrublet"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://github.com/swolock/scrublet" class="uri">https://github.com/swolock/scrublet</a></p>
<p>A python based tool for doublet detection. One of many tools in this space.</p>
</div>
<div id="scvelo" class="section level3 unnumbered">
<h3>ScVelo<a class="anchor" aria-label="anchor" href="#scvelo"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://scvelo.readthedocs.io/" class="uri">https://scvelo.readthedocs.io/</a></p>
<p>A package for single cell RNA velocity analysis, useful for developmental/pseudotime trajectories. Python/scanpy based.</p>
</div>
<div id="monocle" class="section level3 unnumbered">
<h3>Monocle<a class="anchor" aria-label="anchor" href="#monocle"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://cole-trapnell-lab.github.io/monocle3/" class="uri">https://cole-trapnell-lab.github.io/monocle3/</a></p>
<p>A package for single cell developmental//pseudotime trajectory analysis.</p>
</div>
<div id="tidyseurat" class="section level3 unnumbered">
<h3>TidySeurat<a class="anchor" aria-label="anchor" href="#tidyseurat"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://stemangiola.github.io/tidyseurat/" class="uri">https://stemangiola.github.io/tidyseurat/</a></p>
<p>For fans of tidyverse-everything, there’s tidyseurat. Example workflow <a href="https://tidytranscriptomics-workshops.github.io/bioc2022_tidytranscriptomics/articles/tidytranscriptomics_case_study.html">here</a></p>
<hr>
</div>
</div>
<div id="preprocessing-tools" class="section level2" number="13.4">
<h2>
<span class="header-section-number">13.4</span> Preprocessing Tools<a class="anchor" aria-label="anchor" href="#preprocessing-tools"><i class="fas fa-link"></i></a>
</h2>
<p>Tooks that process raw sequencing data into counts matricies</p>
<div id="cell-ranger" class="section level3 unnumbered">
<h3>Cell Ranger<a class="anchor" aria-label="anchor" href="#cell-ranger"><i class="fas fa-link"></i></a>
</h3>
<p><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" class="uri">https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger</a></p>
<p>CellRanger is the 10X tool that takes raw fastq sequence files and produces the counts matricies that are the starting point for today’s analysis. It only works for 10X data.</p>
</div>
<div id="starsolo" class="section level3 unnumbered">
<h3>STARSolo<a class="anchor" aria-label="anchor" href="#starsolo"><i class="fas fa-link"></i></a>
</h3>
<p>STAR is an aligner (which is actually used within cell ranger). STARSolo is a tool for producing counts matricies, and is configurable enough for use with multiple technologies.</p>
<p><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" class="uri">https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md</a></p>
<!-- tidyverse -->
<!-- singleR -->
<!-- celldex -->
<!-- library(SingleCellExperiment) -->
<!-- library(SingleR) -->
<!-- library(celldex) -->
<!-- For monocle: (not done.) -->
<!-- remotes::install_github('satijalab/seurat-wrappers') -->
<!-- onfigure: error: gdal-config not found or not executable. -->
<!-- remotes::install_github('rspatial/terra') -->
<!-- devtools::install_github('cole-trapnell-lab/monocle3') -->

</div>
</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="finishing-up.html"><span class="header-section-number">12</span> Finishing up</a></div>
<div class="next"><a href="references.html">References</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#resources"><span class="header-section-number">13</span> Resources</a></li>
<li>
<a class="nav-link" href="#help-and-fruther-resources"><span class="header-section-number">13.1</span> Help and fruther Resources</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#seurat-vignettes">Seurat Vignettes</a></li>
<li><a class="nav-link" href="#seurat-cheatsheet">Seurat Cheatsheet</a></li>
<li><a class="nav-link" href="#osca">OSCA</a></li>
<li><a class="nav-link" href="#mbp-training-reading-list">MBP training Reading list</a></li>
<li><a class="nav-link" href="#biocommons-single-cell-omics">Biocommons Single Cell Omics</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#data"><span class="header-section-number">13.2</span> Data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#demo-10x-data">Demo 10X data</a></li>
<li><a class="nav-link" href="#geo">GEO</a></li>
<li><a class="nav-link" href="#seurat-data">Seurat data</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#analysis-tools"><span class="header-section-number">13.3</span> Analysis Tools</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#cyclone">Cyclone</a></li>
<li><a class="nav-link" href="#harmony">Harmony</a></li>
<li><a class="nav-link" href="#singler-1">SingleR</a></li>
<li><a class="nav-link" href="#scrublet">Scrublet</a></li>
<li><a class="nav-link" href="#scvelo">ScVelo</a></li>
<li><a class="nav-link" href="#monocle">Monocle</a></li>
<li><a class="nav-link" href="#tidyseurat">TidySeurat</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#preprocessing-tools"><span class="header-section-number">13.4</span> Preprocessing Tools</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#cell-ranger">Cell Ranger</a></li>
<li><a class="nav-link" href="#starsolo">STARSolo</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/06-resources.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/06-resources.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2022-08-18.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
