<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 6 SingleR | scRNAseq Analysis in R with Seurat</title>
<meta name="description" content='#install.packages("BiocManager") #BiocManager::install(c("SingleCellExperiment","SingleR","celldex"),ask=F) library(SingleCellExperiment) library(SingleR) library(celldex) In this workshop we have...'>
<meta name="generator" content="bookdown 0.35 with bs4_book()">
<meta property="og:title" content="Chapter 6 SingleR | scRNAseq Analysis in R with Seurat">
<meta property="og:type" content="book">
<meta property="og:description" content='#install.packages("BiocManager") #BiocManager::install(c("SingleCellExperiment","SingleR","celldex"),ask=F) library(SingleCellExperiment) library(SingleR) library(celldex) In this workshop we have...'>
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 6 SingleR | scRNAseq Analysis in R with Seurat">
<meta name="twitter:description" content='#install.packages("BiocManager") #BiocManager::install(c("SingleCellExperiment","SingleR","celldex"),ask=F) library(SingleCellExperiment) library(SingleR) library(celldex) In this workshop we have...'>
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.5.0/transition.js"></script><script src="libs/bs3compat-0.5.0/tabs.js"></script><script src="libs/bs3compat-0.5.0/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.6.2/htmlwidgets.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.28/datatables.js"></script><link href="libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">scRNAseq Analysis in R with Seurat</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="schedule.html"><span class="header-section-number">2</span> Schedule</a></li>
<li class="book-part">Seurat Preprocessing</li>
<li><a class="" href="preprocessing.html"><span class="header-section-number">3</span> Seurat PBMC3k Tutorial</a></li>
<li><a class="" href="clustering.html"><span class="header-section-number">4</span> Clustering</a></li>
<li><a class="" href="clustermarkers.html"><span class="header-section-number">5</span> Cluster Markers</a></li>
<li class="book-part">Futher Analysis</li>
<li><a class="active" href="singler.html"><span class="header-section-number">6</span> SingleR</a></li>
<li><a class="" href="de2.html"><span class="header-section-number">7</span> Differential Expression</a></li>
<li><a class="" href="CellCycle.html"><span class="header-section-number">8</span> Cell cycle Assignment</a></li>
<li><a class="" href="Harmony.html"><span class="header-section-number">9</span> Data set integration with Harmony</a></li>
<li><a class="" href="resources.html"><span class="header-section-number">10</span> Resources</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/rstudio/bookdown-demo">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="singler" class="section level1" number="6">
<h1>
<span class="header-section-number">6</span> SingleR<a class="anchor" aria-label="anchor" href="#singler"><i class="fas fa-link"></i></a>
</h1>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#install.packages("BiocManager")</span></span>
<span><span class="co">#BiocManager::install(c("SingleCellExperiment","SingleR","celldex"),ask=F)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/SingleR">SingleR</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LTLA/celldex">celldex</a></span><span class="op">)</span></span></code></pre></div>
<p>In this workshop we have focused on the Seurat package. However, there is another whole ecosystem of R packages for single cell analysis within Bioconductor. We won’t go into any detail on these packages in this workshop, but there is good material describing the object type online : <a href="https://robertamezquita.github.io/orchestratingSingleCellAnalysis/data-infrastructure.html">OSCA</a>.</p>
<p>For now, we’ll just convert our Seurat object into an object called SingleCellExperiment. Some popular packages from Bioconductor that work with this type are Slingshot, Scran, Scater.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/as.SingleCellExperiment.html">as.SingleCellExperiment</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></span>
<span><span class="va">sce</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 13714 2638 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts logcounts scaledata</span></span>
<span><span class="co">#&gt; rownames(13714): AL627309.1 AP006222.2 ... PNRC2.1</span></span>
<span><span class="co">#&gt;   SRSF10.1</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(2638): AAACATACAACCAC-1 AAACATTGAGCTAC-1 ...</span></span>
<span><span class="co">#&gt;   TTTGCATGAGAGGC-1 TTTGCATGCCTCAC-1</span></span>
<span><span class="co">#&gt; colData names(31): orig.ident nCount_RNA ...</span></span>
<span><span class="co">#&gt;   pretty_cluster_labels ident</span></span>
<span><span class="co">#&gt; reducedDimNames(2): PCA UMAP</span></span>
<span><span class="co">#&gt; mainExpName: RNA</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
<p>We will now use a package called SingleR to label each cell. SingleR uses a reference data set of cell types with expression data to infer the best label for each cell. A convenient collection of cell type reference is in the <code>celldex</code> package which currently contains the follow sets:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span><span class="op">(</span><span class="st">'package:celldex'</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "BlueprintEncodeData"             </span></span>
<span><span class="co">#&gt; [2] "DatabaseImmuneCellExpressionData"</span></span>
<span><span class="co">#&gt; [3] "HumanPrimaryCellAtlasData"       </span></span>
<span><span class="co">#&gt; [4] "ImmGenData"                      </span></span>
<span><span class="co">#&gt; [5] "MonacoImmuneData"                </span></span>
<span><span class="co">#&gt; [6] "MouseRNAseqData"                 </span></span>
<span><span class="co">#&gt; [7] "NovershternHematopoieticData"</span></span></code></pre></div>
<p>In this example, we’ll use the <code>HumanPrimaryCellAtlasData</code> set, which contains high-level, and fine-grained label types. Lets download the reference dataset</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This too is a sce object,</span></span>
<span><span class="co"># colData is equivalent to seurat's metadata</span></span>
<span><span class="va">ref.set</span> <span class="op">&lt;-</span> <span class="fu">celldex</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/celldex/man/HumanPrimaryCellAtlasData.html">HumanPrimaryCellAtlasData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; snapshotDate(): 2023-04-24</span></span>
<span><span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; see ?celldex and browseVignettes('celldex') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span></code></pre></div>
<p>The “main” labels.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">ref.set</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "DC"                   "Smooth_muscle_cells" </span></span>
<span><span class="co">#&gt;  [3] "Epithelial_cells"     "B_cell"              </span></span>
<span><span class="co">#&gt;  [5] "Neutrophils"          "T_cells"             </span></span>
<span><span class="co">#&gt;  [7] "Monocyte"             "Erythroblast"        </span></span>
<span><span class="co">#&gt;  [9] "BM &amp; Prog."           "Endothelial_cells"   </span></span>
<span><span class="co">#&gt; [11] "Gametocytes"          "Neurons"             </span></span>
<span><span class="co">#&gt; [13] "Keratinocytes"        "HSC_-G-CSF"          </span></span>
<span><span class="co">#&gt; [15] "Macrophage"           "NK_cell"             </span></span>
<span><span class="co">#&gt; [17] "Embryonic_stem_cells" "Tissue_stem_cells"   </span></span>
<span><span class="co">#&gt; [19] "Chondrocytes"         "Osteoblasts"         </span></span>
<span><span class="co">#&gt; [21] "BM"                   "Platelets"           </span></span>
<span><span class="co">#&gt; [23] "Fibroblasts"          "iPS_cells"           </span></span>
<span><span class="co">#&gt; [25] "Hepatocytes"          "MSC"                 </span></span>
<span><span class="co">#&gt; [27] "Neuroepithelial_cell" "Astrocyte"           </span></span>
<span><span class="co">#&gt; [29] "HSC_CD34+"            "CMP"                 </span></span>
<span><span class="co">#&gt; [31] "GMP"                  "MEP"                 </span></span>
<span><span class="co">#&gt; [33] "Myelocyte"            "Pre-B_cell_CD34-"    </span></span>
<span><span class="co">#&gt; [35] "Pro-B_cell_CD34+"     "Pro-Myelocyte"</span></span></code></pre></div>
<p>An example of the types of “fine” labels.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">ref.set</span><span class="op">$</span><span class="va">label.fine</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "DC:monocyte-derived:immature"       </span></span>
<span><span class="co">#&gt; [2] "DC:monocyte-derived:Galectin-1"     </span></span>
<span><span class="co">#&gt; [3] "DC:monocyte-derived:LPS"            </span></span>
<span><span class="co">#&gt; [4] "DC:monocyte-derived"                </span></span>
<span><span class="co">#&gt; [5] "Smooth_muscle_cells:bronchial:vit_D"</span></span>
<span><span class="co">#&gt; [6] "Smooth_muscle_cells:bronchial"</span></span></code></pre></div>
<p>Now we’ll label our cells using the SingleCellExperiment object, with the above reference set.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pred.cnts</span> <span class="op">&lt;-</span> <span class="fu">SingleR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html">SingleR</a></span><span class="op">(</span>test <span class="op">=</span> <span class="va">sce</span>, ref <span class="op">=</span> <span class="va">ref.set</span>, labels <span class="op">=</span> <span class="va">ref.set</span><span class="op">$</span><span class="va">label.main</span><span class="op">)</span></span></code></pre></div>
<p>Keep any types that have more than 10 cells to the label, and put those labels back on our Seurat object and plot our on our umap.</p>
<!-- previously this used first.lables, which no longer seems to exist in pred.cnts -->
<!-- labels and pruned labels are identical in this data -->
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lbls.keep</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">pred.cnts</span><span class="op">$</span><span class="va">labels</span><span class="op">)</span><span class="op">&gt;</span><span class="fl">10</span></span>
<span><span class="va">pbmc</span><span class="op">$</span><span class="va">SingleR.labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="va">lbls.keep</span><span class="op">[</span><span class="va">pred.cnts</span><span class="op">$</span><span class="va">labels</span><span class="op">]</span>, <span class="va">pred.cnts</span><span class="op">$</span><span class="va">labels</span>, <span class="st">'Other'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction<span class="op">=</span><span class="st">'umap'</span>, group.by<span class="op">=</span><span class="st">'SingleR.labels'</span><span class="op">)</span></span></code></pre></div>
<div class="inline-figure"><img src="scRNAseqInR_Doco_files/figure-html/unnamed-chunk-29-1.png" width="672"></div>
<p>It is nice to see that even though SingleR does not use the clusters we computed earlier, the labels do seem to match those clusters reasonably well.</p>

</div>
  <div class="chapter-nav">
<div class="prev"><a href="clustermarkers.html"><span class="header-section-number">5</span> Cluster Markers</a></div>
<div class="next"><a href="de2.html"><span class="header-section-number">7</span> Differential Expression</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav"><li><a class="nav-link" href="#singler"><span class="header-section-number">6</span> SingleR</a></li></ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/rstudio/bookdown-demo/blob/master/04-08-singleR.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/rstudio/bookdown-demo/edit/master/04-08-singleR.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>scRNAseq Analysis in R with Seurat</strong>" was written by . It was last built on 2024-05-13.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
